File: src/traversability/python/polarize.py
"""Cartesian-to-polar stage for offline traversability pipeline."""

from __future__ import annotations

from typing import Any, Dict

import numpy as np


def polarize(points: np.ndarray, config: Dict[str, Any]) -> np.ndarray:
    """Convert XYZ points to [r, theta, z] with z/r filtering.

    Args:
        points: (N, 3) float array [x, y, z].
        config: Polarization config mapping with:
            - z_threshold
            - min_range

    Returns:
        (M, 3) float32 array [r, theta, z].
    """
    points = np.asarray(points)
    if points.ndim != 2 or points.shape[1] < 3:
        raise ValueError(f"`points` must be 2D with at least 3 columns, got {points.shape}")

    xyz = np.asarray(points[:, :3], dtype=np.float32)
    if xyz.size == 0:
        return np.empty((0, 3), dtype=np.float32)

    z_threshold = float(config.get("z_threshold", 1.5))
    min_range = float(config.get("min_range", 0.1))

    if z_threshold < 0.0:
        raise ValueError("`z_threshold` must be >= 0.")
    if min_range < 0.0:
        raise ValueError("`min_range` must be >= 0.")

    x = xyz[:, 0]
    y = xyz[:, 1]
    z = xyz[:, 2]

    z_mask = np.abs(z) < z_threshold
    x = x[z_mask]
    y = y[z_mask]
    z = z[z_mask]

    if x.size == 0:
        return np.empty((0, 3), dtype=np.float32)

    r = np.sqrt(x * x + y * y).astype(np.float32, copy=False)
    theta = np.arctan2(y, x)

    range_mask = r > min_range
    if not np.any(range_mask):
        return np.empty((0, 3), dtype=np.float32)

    polarized = np.column_stack((r[range_mask], theta[range_mask], z[range_mask]))
    return polarized.astype(np.float32, copy=False)

File: src/traversability/polarize.cpp
#include "traversability/polarize.hpp"

#include <cmath>
#include <stdexcept>
#include <string>
#include <vector>

namespace traversability {

Polarize::Polarize(const PolarizeConfig& config) : config_(config) {}

Eigen::MatrixXf Polarize::process(const Eigen::MatrixXf& points) const {
    if (points.cols() < 3) {
        throw std::invalid_argument(
            "`points` must be 2D with at least 3 columns, got cols=" +
            std::to_string(points.cols()));
    }

    const Eigen::MatrixXf xyz = points.leftCols(3);
    if (xyz.size() == 0) {
        return Eigen::MatrixXf(0, 3);
    }

    const float z_threshold = config_.z_threshold;
    const float min_range = config_.min_range;

    if (z_threshold < 0.0f) {
        throw std::invalid_argument("`z_threshold` must be >= 0.");
    }
    if (min_range < 0.0f) {
        throw std::invalid_argument("`min_range` must be >= 0.");
    }

    std::vector<Eigen::Vector3f> z_filtered;
    z_filtered.reserve(static_cast<std::size_t>(xyz.rows()));
    for (Eigen::Index i = 0; i < xyz.rows(); ++i) {
        const float x = xyz(i, 0);
        const float y = xyz(i, 1);
        const float z = xyz(i, 2);
        if (std::abs(z) < z_threshold) {
            z_filtered.emplace_back(x, y, z);
        }
    }

    if (z_filtered.empty()) {
        return Eigen::MatrixXf(0, 3);
    }

    std::vector<Eigen::Vector3f> polar_filtered;
    polar_filtered.reserve(z_filtered.size());
    for (const Eigen::Vector3f& p : z_filtered) {
        const float x = p.x();
        const float y = p.y();
        const float z = p.z();
        const float r = std::sqrt(x * x + y * y);
        const float theta = std::atan2(y, x);

        if (r > min_range) {
            polar_filtered.emplace_back(r, theta, z);
        }
    }

    if (polar_filtered.empty()) {
        return Eigen::MatrixXf(0, 3);
    }

    Eigen::MatrixXf out(static_cast<Eigen::Index>(polar_filtered.size()), 3);
    for (Eigen::Index i = 0; i < static_cast<Eigen::Index>(polar_filtered.size()); ++i) {
        out.row(i) = polar_filtered[static_cast<std::size_t>(i)];
    }
    return out;
}

} // namespace traversability
